### ISSUE TITLE
Can't get custom error messages to work.

### ISSUE THREAD

**Issue by rowlesmr** [taocpp/PEGTL#353]
I am getting a static assert error: `error C2338: static_assert failed: 'explicit error message required for Rule'` (from line 61 of tao\pegtl\must_if.hpp). I don't understand what I need to do to satisfy the assert, or even when the error means.

If I run the [error message](https://github.com/taocpp/PEGTL/blob/788f223ee7c8377f78966c87d60b91188e3df33a/src/test/pegtl/error_message.cpp) example, it works, so I know my setup can allow it to happen.

.

As an example, I have an input which fails with the error message:

```
C:\_______\data\test_data\test.cif:6:23: parse error matching struct row::cif::rules::lst_end
_list2 [ val val2 val3 ]
                      ^
```

I would like to specialise the error message.

From the examples, I've put together:

```C++
// I'm in the namespace row::cif
template< typename > inline constexpr const char* error_message = nullptr;
template<> inline constexpr auto error_message< rules::lst_end > = "Incorrect list ending.";

struct error {
    template< typename Rule > static constexpr auto message = error_message< Rule >;
};

template< typename Rule >
using control = pegtl::must_if< error >::control< Rule >;
```

and then called the parse function as

```C++
// I'm in the row::cif namespace
pegtl::parse < rules::CIF2_file, pegtl::nothing, control> (in);
```

I can't see anything obviously wrong with what I've done, unless I've put an error message on something that shouldn't, or need to put a message on something else.

.
I have to see if I can make up a mwe without breaking anything else.

**Comment by d-frey** [comment]
I can't see anything obviously wrong, either. Can you reduce your code into a small but complete example and post it?

**Comment by d-frey** [comment]
I think the issue is that you have more rules to cover. Your grammar uses:

```
if_must< lst_begin, lst_values, lst_end >
```

meaning that you need to provide error message specializations for `lst_end`, but also for `lst_values`. Also, you grammar contains

```
if_must<one<'#'>, until<eolf>>
```

which means you also need to specialize for `until<eolf>`. Once I add the two missing specializations, it start to compile. https://godbolt.org/z/fojoh8159

With other compilers, the error messages from the compiler also contain `lst_values` and `until<eolf>` to hint at which specializations are missing. I haven't tested it, but I think MSVC should also do this. But well... it's still C++ error messages and the PEGTL is quite heavy on templates, so it's definitely not easy to spot it immediately. ;)

**Comment by rowlesmr** [comment]
Knowing the answer, I can now kind of parse the error messages. I can also see how I read the docs wrong. I'll have to see how I'd reword it.

**Comment by rowlesmr** [comment]
>`until<eolf>`

this is why it failed when I went nuclear and added all of my rules; I didn't add that one.

**Comment by d-frey** [comment]
> > `until<eolf>`
> 
> this is why it failed when I went nuclear and added all of my rules; I didn't add that one.

Note that you probably don't want to specialize for `until<eolf>` directly, as you might have this rule in several places in the future within your grammar. You would want to add `struct comment_end : until<eolf> {};`, use that in your `comment` rule and specialize for `comment_end`.

**Comment by rowlesmr** [comment]
> You would want to add `struct comment_end : until<eolf> {};`, use that in your `comment` rule and specialize for `comment_end`.

Wilco.

**Comment by rowlesmr** [comment]
Can confirm. Adding those missing rules enabled compilation.
